### PR TITLE
Pre & Post ClassVisitor

### PR DESCRIPTION
**Pull request by Devan-Kerman**
a pre & post-class visitor, along with a small API

### CODE REVIEW

**Comment by Devan-Kerman** [comment]
need to fix some compile errors I think, merge probably wasn't perfect

**Comment by modmuss50** [comment]
What is the usecase for this?

**Comment by Devan-Kerman** [comment]
remapping mixins and other bits

**Comment by modmuss50** [comment]
Oh i see, for somereason I throught this was the access widener repo, opps...

This is one for @sfPlayer1 to look at then.

File: src/main/java/net/fabricmc/tinyremapper/ClassInstance.java [hunk 1]
```diff
@@ -652,6 +705,7 @@ public static String getMrjName(String clsName, int mrjVersion) {
 	private String name;
 	private int mrjVersion;
 	private String superName;
+	private String signature;
```
Comment (LogicFan) [inline src/main/java/net/fabricmc/tinyremapper/ClassInstance.java:708]: What is the difference between signature and name?

File: src/main/java/net/fabricmc/tinyremapper/TinyRemapper.java [hunk 2]
```diff
@@ -161,14 +164,24 @@ public Builder extraRemapper(Remapper remapper) {
 			return this;
 		}
 
+		public Builder extraPreVisitor(WrapperFunction func) {
+			this.pre = func;
+			return this;
+		}
+
+		public Builder extraPostVisitor(WrapperFunction function) {
```
Comment (LogicFan) [inline src/main/java/net/fabricmc/tinyremapper/TinyRemapper.java:172]: parameter name should be consistent.

**Comment by LogicFan** [comment]
I am interested in using this new feature to add to the ability to remap mixin inside tiny-remapper. Do you have any idea how that can be done?

File: src/main/java/net/fabricmc/tinyremapper/api/WrapperFunction.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package net.fabricmc.tinyremapper.api;
+
+import org.objectweb.asm.ClassVisitor;
+
+public interface WrapperFunction {
```
Comment (LogicFan) [inline src/main/java/net/fabricmc/tinyremapper/api/WrapperFunction.java:5]: Maybe add documentation on how this one should be used.

File: src/main/java/net/fabricmc/tinyremapper/TinyRemapper.java [hunk 4]
```diff
@@ -206,11 +220,13 @@ private TinyRemapper(Collection<IMappingProvider> mappingProviders, boolean igno
...
 			boolean skipLocalMapping,
 			boolean renameInvalidLocals,
-			ClassVisitor extraAnalyzeVisitor, Remapper extraRemapper) {
+			ClassVisitor extraAnalyzeVisitor, Remapper extraRemapper, WrapperFunction pre, WrapperFunction post) {
 		this.mappingProviders = mappingProviders;
 		this.ignoreFieldDesc = ignoreFieldDesc;
 		this.threadCount = threadCount > 0 ? threadCount : Math.max(Runtime.getRuntime().availableProcessors(), 2);
 		this.keepInputData = keepInputData;
+		this.pre = pre;
+		this.post = post;
```
Comment (LogicFan) [inline src/main/java/net/fabricmc/tinyremapper/TinyRemapper.java:229]: Why would you only allow one pre/post visitor. Imagine that there could be different pre/post visitors for different purposes and written by different people, they should all be able to apply in sequence. For example, it should allow `builder.extraPre(...).extraPre(...).extraPre(...)` and apply them in sequence.

Comment (Chocohead) [reply src/main/java/net/fabricmc/tinyremapper/TinyRemapper.java:229]: `ClassVisitor`s naturally chain, so if you had multiple you wanted to apply you would simply chain them in your `WrapperFunction`

File: src/main/java/net/fabricmc/tinyremapper/ClassInstance.java (hunk 1 above)
Comment (Chocohead) [reply src/main/java/net/fabricmc/tinyremapper/ClassInstance.java:708]: The name is the internal name of the class, signature is the generic signature of the class (which may be `null` if the class isn't parametrised)

File: src/main/java/net/fabricmc/tinyremapper/AsmClassRemapper.java [hunk 5]
```diff
@@ -113,17 +115,19 @@ public MethodVisitor visitMethod(int access, String name, String descriptor, Str
 			methodNode = new MethodNode(api, access, name, descriptor, signature, exceptions);
 		}
 
-		return super.visitMethod(access, name, descriptor, signature, exceptions);
+		MethodVisitor mv = super.visitMethod(access, name, descriptor, signature, exceptions);
+		MemberHeader header = null;
```
Comment (Chocohead) [inline src/main/java/net/fabricmc/tinyremapper/AsmClassRemapper.java:119]: Should this be `null`?

File: src/main/java/net/fabricmc/tinyremapper/AsmClassRemapper.java [hunk 6]
```diff
@@ -96,11 +96,13 @@ public void visitSource(String source, String debug) {
 
 	@Override
 	public FieldVisitor visitField(int access, String name, String descriptor, String signature, Object value) {
+		MemberHeader header = null;
```
Comment (Chocohead) [inline src/main/java/net/fabricmc/tinyremapper/AsmClassRemapper.java:99]: Should this be `null`?

File: build.gradle [hunk 7]
```diff
@@ -18,6 +18,10 @@ archivesBaseName = 'tiny-remapper'
 
 repositories {
 	mavenCentral()
+	maven {
+		name = 'fabric'
+		url = 'https://maven.fabricmc.net/'
+	}
```
Comment (Chocohead) [inline build.gradle:24]: Suppose we don't need this now Mixin isn't a dependency

File: src/main/java/net/fabricmc/tinyremapper/ClassInstance.java [hunk 8]
```diff
@@ -148,10 +163,48 @@ public int getMrjVersion() {
...
 	}
 
+	@Override
+	public String getSignature() {
+		return signature;
+	}
+
+	@Override
+	public MemberHeader getField(String name, String desc) {
+		MemberInstance instance = this.resolvedMembers.get(MemberInstance.getFieldId(name, desc, false));
```
Comment (Chocohead) [inline src/main/java/net/fabricmc/tinyremapper/ClassInstance.java:178]: Could allow `null` field descriptors if `false` is switched to `desc == null`

File: src/main/java/net/fabricmc/tinyremapper/TinyRemapper.java (hunk 4 above)
Comment (LogicFan) [reply src/main/java/net/fabricmc/tinyremapper/TinyRemapper.java:229]: But then the function name should not be called `extraPre` or `extraPost` as it confuses people. It should be `setPre` or `setPost` since the "extra" indicates you can have them chain together.

Comment (Devan-Kerman) [reply src/main/java/net/fabricmc/tinyremapper/TinyRemapper.java:229]: the other methods don't follow that convention

**Review by LogicFan — APPROVED** [review]

File: src/main/java/net/fabricmc/tinyremapper/api/WrapperFunction.java (hunk 3 above)
Comment (LogicFan) [reply src/main/java/net/fabricmc/tinyremapper/api/WrapperFunction.java:8]: Maybe also consider adding `@FunctionalInterface` for this interface.

File: src/main/java/net/fabricmc/tinyremapper/api/WrapperFunction.java [hunk 9]
```diff
@@ -0,0 +1,10 @@
+package net.fabricmc.tinyremapper.api;
+
+import org.objectweb.asm.ClassVisitor;
+
+/**
+ * @see ClassVisitor#ClassVisitor(int, ClassVisitor)
+ */
+public interface WrapperFunction {
+	ClassVisitor wrap(ClassVisitor visitor, ExtendedRemapper remapper, Classpath classpath);
```
Comment (LogicFan) [inline src/main/java/net/fabricmc/tinyremapper/api/WrapperFunction.java:9]: Maybe also consider adding @FunctionalInterface for this interface.

File: src/main/java/net/fabricmc/tinyremapper/TinyRemapper.java [hunk 10]
```diff
@@ -956,11 +971,21 @@ private void mrjRefresh(MrjState state) {
...
-		reader.accept(new AsmClassRemapper(visitor, cls.getContext().remapper, rebuildSourceFilenames, checkPackageAccess, skipLocalMapping, renameInvalidLocals), flags);
+		if (post != null) {
+			visitor = post.wrap(visitor, this.getRemapper(), this);
+		}
+
+		visitor = new AsmClassRemapper(visitor, cls.getContext().remapper, rebuildSourceFilenames, checkPackageAccess, skipLocalMapping, renameInvalidLocals);
+
+		if (pre != null) {
+			visitor = pre.wrap(visitor, this.getRemapper(), this);
+		}
```
Comment (LogicFan) [inline src/main/java/net/fabricmc/tinyremapper/TinyRemapper.java:985]: the `this.getRemapper()` does not respect with the mrj version. Shouldn't this be `cls.getContext.remapper`?

Comment (Chocohead) [reply src/main/java/net/fabricmc/tinyremapper/TinyRemapper.java:985]: This PR was made before the MRJ PR was merged so this is probably a merging oversight

**Comment by Chocohead** [comment]
Are there any changes to how `AsmClassRemapper` works now? The only remaining changes look either to be stylistic or functionally no different to how it previously worked (`createMethodRemapper` and `createFieldRemapper`)

File: src/main/java/net/fabricmc/tinyremapper/MetaInfFixer.java [hunk 11]
```diff
@@ -39,7 +39,7 @@ public void transform(Path destinationDirectory, Path relativePath, InputStream
 			Path outputDir = outputFile.getParent();
 			if (outputDir != null) Files.createDirectories(outputDir);
 
-			try (OutputStream os = new BufferedOutputStream(Files.newOutputStream(outputFile))) {
+			try (OutputStream os = new BufferedOutputStream(Files.newOutputStream(outputFile, StandardOpenOption.CREATE))) {
```
Comment (Chocohead) [inline src/main/java/net/fabricmc/tinyremapper/MetaInfFixer.java:42]: The default options (having passed none) is `StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING, StandardOpenOption.WRITE`, is this for something like avoiding overwriting service files?

Comment (Devan-Kerman) [reply src/main/java/net/fabricmc/tinyremapper/MetaInfFixer.java:42]: I think that's how it was originally, maybe it was a bad merge

File: src/main/java/net/fabricmc/tinyremapper/api/MemberHeader.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
+package net.fabricmc.tinyremapper.api;
+
+import java.util.Objects;
+
+public class MemberHeader {
+	public final ClassHeader header;
```
Comment (liach) [inline src/main/java/net/fabricmc/tinyremapper/api/MemberHeader.java:6]: Probably call this field `owner` instead. `header.header` is quite confusing

File: src/main/java/net/fabricmc/tinyremapper/api/WrapperFunction.java (hunk 9 above)
Comment (liach) [reply src/main/java/net/fabricmc/tinyremapper/api/WrapperFunction.java:10]: Must the visitor propagate at real time or can it cache the results and mass dump calls later? Such as we first grab everything from the source visitor to a class node, do stuff to the node, and finally we dump a visitor to the node, instead of delegating most calls and only patching a few selected calls.

File: src/main/java/net/fabricmc/tinyremapper/api/WrapperFunction.java (hunk 3 above)
Comment (liach) [reply src/main/java/net/fabricmc/tinyremapper/api/WrapperFunction.java:9]: Should add a simple API like `andThen` to chain wrappers easily, and use that to support multiple wrappers in remapper builder code

**Comment by liach** [comment]
2 comments on `WrapperFunction`:
1. We can just add an `andThen` default method (or static method) to chain pre/post wrappers in the remapper builder
2. For wrapping usage:
Must the visitor propagate at real time or can it cache the results and mass dump calls later? Such as we first grab everything from the source visitor to a class node, do stuff to the node, and finally we dump a visitor to the node, instead of patching a few selected calls and delegating other calls. If this behavior is prohibited, it may be worthy to document.

Thanks github, my 2 comments on desired lines are considered followup comments of the resolved discussion. :vomiting_face:

**Comment by LogicFan** [comment]
> Must the visitor propagate at real time or can it cache the results and mass dump calls later?

Both can be done with the current ASM structure. There is no reason to ban one of them, and both of them have use cases.

**Comment by liach** [comment]
Yeah, that should be fine, but if there is any peculiarities caused by such buffering, I mean it may be worthy to document.

**Comment by LogicFan** [comment]
I think it is also necessary to have a visitor when reads the class; or have the ability to work by round globally. For example, if I want to remap a public `@Shadow` field, it would be necessary to remap the whole thing instead of just a single class. The current API lacks this functionality.

**Comment by LogicFan** [comment]
Why would you want to feed `Classpath` as a parameter? You can get it inside `ClassVisitor`, right?

**Comment by LogicFan** [comment]
Any ETA when this one will be merged? #63 depends on this.

**Comment by Devan-Kerman** [comment]
I think I will add an andThen simply for convienence

**Comment by liach** [comment]
@sfPlayer1 mind review?

File: src/main/java/net/fabricmc/tinyremapper/ClassInstance.java [hunk 13]
```diff
@@ -148,10 +163,48 @@ public int getMrjVersion() {
...
+			return new MemberHeader(this, instance.access, instance.name, instance.desc);
+		}
+
+		return null;
+	}
+
+	@Override
+	public Iterable<MemberHeader> allMembers() {
+		return MappedIterator.map(this.resolvedMembers.values(), i -> new MemberHeader(this, i.access, i.name, i.desc));
+	}
```
Comment (LogicFan) [inline src/main/java/net/fabricmc/tinyremapper/ClassInstance.java:201]: `resolvedMembers` is a cache. It may or may or contains the member from parents' classes, which introduces inconsistency. I think a better way is needed.

File: src/main/java/net/fabricmc/tinyremapper/api/MemberHeader.java [hunk 14]
```diff
@@ -0,0 +1,44 @@
+package net.fabricmc.tinyremapper.api;
+
+import java.util.Objects;
+
+public class MemberHeader {
```
Comment (LogicFan) [inline src/main/java/net/fabricmc/tinyremapper/api/MemberHeader.java:5]: Is there any specific reason to not make this an interface inherited by `MemberInstance`?

**Review by LogicFan — CHANGES_REQUESTED** [review]

**Comment by LogicFan** [comment]
Here is how `resolve` is written in the `ClassInstance`.

```Java
public MemberInstance resolve(MemberType type, String id) {
		MemberInstance member = getMember(type, id);
		if (member != null) return member;

		// get from cache
		member = resolvedMembers.get(id);

		if (member == null) {
			// compute
			member = resolve0(type, id);
			assert member != null;

			// put in cache
			MemberInstance prev = resolvedMembers.putIfAbsent(id, member);
			if (prev != null) member = prev;
		}

		return member != nullMember ? member : null;
	}
```

You can see that `resolvedMemebrs` is a cache that generated when necessary. i.e. the member resolution is lazy. Thus, you cannot directly use `resolvedMembers`.

**Comment by LogicFan** [comment]
My suggestion is to make the `MemberHeader` an interface inherited by `MemberInstance`, and use `resolve` to get the method & field for `getMethod` and `getField`. 

`allMembers` is tricky, either you need to recursively resolve all members from parents, or just remove this method I guess.

**Comment by LogicFan** [comment]
Any update on this PR?

**Comment by Devan-Kerman** [comment]
I was out, am back now, will fix

**Comment by Devan-Kerman** [comment]
everything should be fixed now

File: src/main/java/net/fabricmc/tinyremapper/ClassInstance.java [hunk 15]
```diff
@@ -148,10 +163,37 @@ public int getMrjVersion() {
...
+
+	@Override
+	public MemberHeader getMethod(String name, String desc) {
+		return this.resolve(MemberType.METHOD, MemberInstance.getMethodId(name, desc));
+	}
+
+	@Override
+	@SuppressWarnings("ALL")
+	public Iterable<MemberHeader> allMembers() {
+		return (Iterable) this.members.values();
```
Comment (Chocohead) [inline src/main/java/net/fabricmc/tinyremapper/ClassInstance.java:189]: You could fix the nasty generics here doing `Iterable<? extends MemberHeader>` as the return type

Comment (Devan-Kerman) [reply src/main/java/net/fabricmc/tinyremapper/ClassInstance.java:189]: yea but then the api looks uglier, better to have the ugly in impl than in API

**Comment by Chocohead** [comment]
> 
> 
> Are there any changes to how `AsmClassRemapper` works now? The only remaining changes look either to be stylistic or functionally no different to how it previously worked (`createMethodRemapper` and `createFieldRemapper`)

Think this question got missed in between reviews too

File: src/main/java/net/fabricmc/tinyremapper/AsmClassRemapper.java [hunk 16]
```diff
@@ -100,7 +99,7 @@ public FieldVisitor visitField(int access, String name, String descriptor, Strin
 			PackageAccessChecker.checkDesc(className, descriptor, "field descriptor", (AsmRemapper) remapper);
 		}
 
-		return super.visitField(access, name, descriptor, signature, value);
+		return new AsmFieldRemapper(super.visitField(access, name, descriptor, signature, value), remapper);
 	}
```
Comment (LogicFan) [inline src/main/java/net/fabricmc/tinyremapper/AsmClassRemapper.java:103]: Any reason to change the remap logic here? (And some other places in the AsmClassRemapper).

File: src/main/java/net/fabricmc/tinyremapper/ClassInstance.java (hunk 15 above)
Comment (Chocohead) [reply src/main/java/net/fabricmc/tinyremapper/ClassInstance.java:189]: It's an `Iterable` though, your normal use in a `for` loop wouldn't be any different. If it was returning `Collection` directly then I'd understand, as it would as it would stop you adding to it (not that you'd be able to in this particular instance anyway), but here it is just using generics as they were intended

Comment (LogicFan) [reply src/main/java/net/fabricmc/tinyremapper/ClassInstance.java:189]: There are two things I want to comment,
1. I really don't think `allMembers` is a good name since it does not retrieve all members, it only retrieves the members defined in this class, not its interface or superclass.
2. Why use iterable, not collection. I personally think the collection is easier to work with than iterable.

**Comment by sfPlayer1** [comment]
What about renaming *Header to Tr*?

File: src/main/java/net/fabricmc/tinyremapper/ClassInstance.java [hunk 17]
```diff
@@ -148,10 +163,37 @@ public int getMrjVersion() {
...
+
+	@Override
+	public MemberHeader getField(String name, String desc) {
+		return this.resolve(MemberType.FIELD, MemberInstance.getFieldId(name, desc, desc == null));
+	}
+
+	@Override
+	public MemberHeader getMethod(String name, String desc) {
+		return this.resolve(MemberType.METHOD, MemberInstance.getMethodId(name, desc));
+	}
```
Comment (LogicFan) [inline src/main/java/net/fabricmc/tinyremapper/ClassInstance.java:184]: Can you also have a version of `getField` and `getMethod` that does not require desc? Because my implementation in #63 would require this capability. You can use `resolvePartial` for this purpose.

**Review by LogicFan — CHANGES_REQUESTED** [review]

**Comment by LogicFan** [comment]
> What about renaming `*Header` to `Tr*`?

This will break lots of code in #63 as it starts uses this API already. Any specific reason you want to make this change?

**Comment by sfPlayer1** [comment]
> > What about renaming `*Header` to `Tr*`?
> 
> This will break lots of code in #63 as it starts uses this API already. Any specific reason you want to make this change?

The Header suffix makes no sense to me, better to use Tr (for tiny remapper) for such a special purpose API. It may be quite easy to merge into your PR if you squash everything, do the same refactors as here, then rebase onto the merged #62.

File: src/main/java/net/fabricmc/tinyremapper/api/ClassHeader.java [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+	/**
+	 * get field by name and descriptor, this is not cached.
+	 * @param desc can be null
+	 */
+	MemberHeader getField(String name, String desc);
+
+	/**
+	 * get method by name and descriptor, this is not cached.
+	 */
+	MemberHeader getMethod(String name, String desc);
```
Comment (LogicFan) [inline src/main/java/net/fabricmc/tinyremapper/api/ClassHeader.java:27]: Since you use `resolve`, which uses the exact match. The documentation here is not correct anymore.

File: src/main/java/net/fabricmc/tinyremapper/ClassInstance.java (hunk 15 above)
Comment (LogicFan) [reply src/main/java/net/fabricmc/tinyremapper/ClassInstance.java:189]: >  your normal use in a for loop wouldn't be any different.

I usually work with `Stream`. Can `Iterable` easily converted to stream? Sorry for the stupid question, I really never touched iterable before.

File: src/main/java/net/fabricmc/tinyremapper/api/WrapperFunction.java [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+
+			return current;
+		};
+	}
+
+	ClassVisitor wrap(ClassVisitor visitor, ExtendedRemapper remapper, Classpath classpath);
+
+	default WrapperFunction andThen(WrapperFunction target) {
+		return (visitor, remapper, classpath) -> target.wrap(this.wrap(visitor, remapper, classpath), remapper, classpath);
+	}
```
Comment (sfPlayer1) [inline src/main/java/net/fabricmc/tinyremapper/api/WrapperFunction.java:32]: Is there any point to the static methods? The API user shouldn't be encouraged to chain wrap or otherwise split the visitors.

**Comment by LogicFan** [comment]
Here is a reference implementation for `getMethod` and `getField` that does not use desc that I used in my PR

```Java
	@Override
	public MemberHeader getField(String name) {
		return this.resolvePartial(MemberType.FIELD, name, ";;");
	}
	@Override
	public MemberHeader getMethod(String name) {
		return this.resolvePartial(MemberType.METHOD, name, "(");
	}
```

File: src/main/java/net/fabricmc/tinyremapper/api/Classpath.java [hunk 20]
```diff
@@ -0,0 +1,10 @@
+package net.fabricmc.tinyremapper.api;
+
+public interface Classpath {
```
Comment (sfPlayer1) [inline src/main/java/net/fabricmc/tinyremapper/api/Classpath.java:3]: The name of this class is a bit weird considering what it does. I call this Environment in Matcher, so maybe `TrEnvironment`? getByName would have to be called getClass or similar then.

File: src/main/java/net/fabricmc/tinyremapper/api/ClassHeader.java [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+
+	int getAccess();
+
+	String getName();
+
+	String getSuperName();
+
+	String getSignature();
+
+	List<String> getInterfaceList();
```
Comment (sfPlayer1) [inline src/main/java/net/fabricmc/tinyremapper/api/ClassHeader.java:16]: getInterfaces? Why is this a List and the other is an Iterable?

File: src/main/java/net/fabricmc/tinyremapper/api/ClassHeader.java [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+	 * @param desc can be null
+	 */
+	MemberHeader getField(String name, String desc);
+
+	/**
+	 * get method by name and descriptor, this is not cached.
+	 */
+	MemberHeader getMethod(String name, String desc);
+
+	Iterable<MemberHeader> allMembers();
```
Comment (sfPlayer1) [inline src/main/java/net/fabricmc/tinyremapper/api/ClassHeader.java:29]: Iterable is usually a bad type to export, Collection is much better.

**Comment by LogicFan** [comment]
> The Header suffix makes no sense to me, better to use Tr (for tiny remapper) for such a special purpose API. It may be quite easy to merge into your PR if you squash everything, do the same refactors as here, then rebase onto the merged #62.

Ya, that would make sense.

File: src/main/java/net/fabricmc/tinyremapper/TinyRemapper.java [hunk 23]
```diff
@@ -1181,7 +1215,12 @@ public void run() {
 		/**
 		 * Propagate names into methods and create additional bridges to keep the normally mapped method name intact.
 		 */
-		COMPATIBLE;
+		COMPATIBLE
+	}
+
+	@Override
+	public ClassHeader getByName(String internalName) {
+		return this.defaultState.classes.get(internalName);
```
Comment (sfPlayer1) [inline src/main/java/net/fabricmc/tinyremapper/TinyRemapper.java:1223]: This isn't really MRJ compatible, apply is MRJ aware, so the environment may be better served by MrjState?.

File: src/main/java/net/fabricmc/tinyremapper/TinyRemapper.java [hunk 24]
```diff
@@ -1142,17 +1178,15 @@ public void run() {
...
-					MemberInstance getter = cls.getMember(MemberType.METHOD, getterIdSrc);
+					MemberInstance getter = cls.getMember(MemberHeader.MemberType.METHOD, getterIdSrc);
 
 					if (getter != null && getter.isVirtual()) {
 						visitedUp.add(cls);
 						visitedDown.add(cls);
-						cls.propagate(TinyRemapper.this, MemberType.METHOD, className, getterIdSrc, nameDst,
-								Direction.ANY, true, true,
-								true, visitedUp, visitedDown);
+						cls.propagate(TinyRemapper.this, MemberHeader.MemberType.METHOD, className, getterIdSrc, nameDst, Direction.ANY, true, true, true, visitedUp, visitedDown);
```
Comment (sfPlayer1) [inline src/main/java/net/fabricmc/tinyremapper/TinyRemapper.java:1189]: Please avoid unrelated changes, the MemberType changes above can be eliminated by changing the import as well.

File: src/main/java/net/fabricmc/tinyremapper/MappedIterator.java [hunk 25]
```diff
@@ -0,0 +1,28 @@
+package net.fabricmc.tinyremapper;
+
+import java.util.Iterator;
+import java.util.function.Function;
+
+public final class MappedIterator<F, T> implements Iterator<T> {
```
Comment (sfPlayer1) [inline src/main/java/net/fabricmc/tinyremapper/MappedIterator.java:6]: What is the point of this? Seems unused?

File: src/main/java/net/fabricmc/tinyremapper/ClassInstance.java (hunk 15 above)
Comment (sfPlayer1) [reply src/main/java/net/fabricmc/tinyremapper/ClassInstance.java:189]: I'd use `Collection<? extends TrMember> getMembers()`

File: src/main/java/net/fabricmc/tinyremapper/ClassInstance.java (hunk 17 above)
Comment (sfPlayer1) [reply src/main/java/net/fabricmc/tinyremapper/ClassInstance.java:184]: I'd call resolving lookup methods resolveField/resolveMethod, compared to getField/getMethod they search super classes/interfaces for the member - not just the class itself.

**Comment by Devan-Kerman** [comment]
There are no intentional changes to AsmRemapper

**Comment by Devan-Kerman** [comment]
Prototype is an alternative name

**Comment by sfPlayer1** [comment]
IMO the overly generic sounding/universal API names should be Tr prefixed, this is a very special purpose library that sits in every mod dev's class path and thus risks messing with auto-imports.

**Comment by Devan-Kerman** [comment]
fair enough

File: src/main/java/net/fabricmc/tinyremapper/api/ClassHeader.java (hunk 21 above)
Comment (Devan-Kerman) [reply src/main/java/net/fabricmc/tinyremapper/api/ClassHeader.java:16]: Order is important here but not for members (I think)

**Comment by Devan-Kerman** [comment]
does TR* work instead of Tr*? TR is an acronym

File: src/main/java/net/fabricmc/tinyremapper/MappedIterator.java (hunk 25 above)
Comment (Devan-Kerman) [reply src/main/java/net/fabricmc/tinyremapper/MappedIterator.java:6]: it's unused now yea, forgot to remove it

File: src/main/java/net/fabricmc/tinyremapper/TinyRemapper.java (hunk 24 above)
Comment (Devan-Kerman) [reply src/main/java/net/fabricmc/tinyremapper/TinyRemapper.java:1189]: this fails checkstyle

**Comment by sfPlayer1** [comment]
abbreviations always count as a single word (Url, Http, Tr, ...)

File: src/main/java/net/fabricmc/tinyremapper/AsmClassRemapper.java (hunk 16 above)
Comment (Devan-Kerman) [reply src/main/java/net/fabricmc/tinyremapper/AsmClassRemapper.java:103]: just passing the remapper instance

Comment (Devan-Kerman) [reply src/main/java/net/fabricmc/tinyremapper/AsmClassRemapper.java:103]: just passing the remapper instance

File: src/main/java/net/fabricmc/tinyremapper/ClassInstance.java (hunk 17 above)
Comment (Devan-Kerman) [reply src/main/java/net/fabricmc/tinyremapper/ClassInstance.java:184]: might be better to make desc nullable

File: src/main/java/net/fabricmc/tinyremapper/ClassInstance.java [hunk 26]
```diff
@@ -174,19 +174,22 @@ public String getSignature() {
 	}
 
 	@Override
-	public MemberHeader getField(String name, String desc) {
+	public TrMember resolveField(String name, String desc) {
 		return this.resolve(MemberType.FIELD, MemberInstance.getFieldId(name, desc, desc == null));
 	}
```
Comment (LogicFan) [inline src/main/java/net/fabricmc/tinyremapper/ClassInstance.java]: The internal implementation uses `tr.ignoreFieldDesc` everywhere. Say if the `tr.ignoreFieldDesc = true` and someone passes `null`, then it will always return no match since the internal id does not match (one id has desc another id does not). Vice versa.

Comment (LogicFan) [reply src/main/java/net/fabricmc/tinyremapper/ClassInstance.java]: Here is code that make things work

```Java
	@Override
	public TrMember resolveField(String name, String desc) {
		if (desc == null) {
			return this.resolvePartial(MemberType.FIELD, name, ";;");
		} else {
			return this.resolve(MemberType.FIELD, MemberInstance.getFieldId(name, desc, tr.ignoreFieldDesc));
		}
	}
```

**Review by LogicFan — APPROVED** [review]
I think this PR is good to go. My only concern is the change of `AsmClassRemapper`, which is the part of the code I do not fully understand. @sfPlayer1 mind to have a look again?

File: src/main/java/net/fabricmc/tinyremapper/AsmClassRemapper.java (hunk 16 above)
Comment (sfPlayer1) [reply src/main/java/net/fabricmc/tinyremapper/AsmClassRemapper.java:103]: I don't get it, the remapper is being passed by the previous implementation as well? It just funneled the remapper instance through the super class field

File: src/main/java/net/fabricmc/tinyremapper/api/TrClass.java [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+
+	int getAccess();
+
+	String getName();
+
+	String getSuperName();
+
+	String getSignature();
+
+	List<String> getInterfaceList();
```
Comment (sfPlayer1) [inline src/main/java/net/fabricmc/tinyremapper/api/TrClass.java:17]: getInterfaces

File: src/main/java/net/fabricmc/tinyremapper/api/TrClass.java [hunk 28]
```diff
@@ -0,0 +1,29 @@
...
+	/**
+	 * @param desc can be null
+	 */
+	TrMember resolveField(String name, String desc);
+
+	/**
+	 */
+	TrMember resolveMethod(String name, String desc);
+
+	Collection<? extends TrMember> allMembers();
```
Comment (sfPlayer1) [inline src/main/java/net/fabricmc/tinyremapper/api/TrClass.java:28]: getMembers

File: src/main/java/net/fabricmc/tinyremapper/api/TrEnvironment.java [hunk 29]
```diff
@@ -0,0 +1,10 @@
+package net.fabricmc.tinyremapper.api;
+
+public interface TrEnvironment {
+	/**
+	 * @return the class with the passed name, or null if not found.
+	 */
+	TrClass getByName(String internalName);
```
Comment (sfPlayer1) [inline src/main/java/net/fabricmc/tinyremapper/api/TrEnvironment.java:7]: getClass

File: src/main/java/net/fabricmc/tinyremapper/api/TrClass.java (hunk 27 above)
Comment (LogicFan) [reply src/main/java/net/fabricmc/tinyremapper/api/TrClass.java:17]: I think it is also better to change the return type of this method to a Collection since the `getMembers` is Collection.

**Comment by LogicFan** [comment]
Any update for this PR?

**Comment by Devan-Kerman** [comment]
should be updated now

**Review by sfPlayer1 — APPROVED** [review]
LGTM now, thanks!
